src/main/java/com/renatusnetwork/momentum/data/menus/MenuItemAction.java

```diff
@@ -646,14 +655,18 @@ public static void performLevelTeleport(PlayerStats playerStats, Level level)
                     }
                 }
 
-                if (playerStats.inLevel() && playerStats.hasAutoSave() && !playerStats.getPlayer().isOnGround())
-                {
+                if (playerStats.inLevel() && playerStats.hasAutoSave() && !playerStats.getPlayer().isOnGround()) {
                     player.sendMessage(Utils.translate("&cYou cannot leave the level while in midair with auto-save enabled"));
                     return;
                 }
 
-                // perform leave level steps
-                statsManager.leaveLevelAndReset(playerStats, true);
+                // perform leave level steps if theyre not toggling mastery of same level
+                if (!enteringMasteryOfSameLevel)
+                    statsManager.leaveLevelAndReset(playerStats, true);
+
+                // autosave so when entering mastery progress gets saved from non-mastery attempt
+                if (enteringMasteryOfSameLevel && playerStats.hasAutoSave())
```

Comment:
can we put this in an else:
if (!enteringMasteryOfSameLevel)
    statsManager.leaveLevelAndReset(playerStats, true);
else if (playerStats.hasAutoSave()
   ... (rest of code)

just a bit better logic since no need to check enteringMasteryOfSameLevel twice

Reply:
changed